ADBDEV-6928: Update limit path creation

ShapatinAA opened the pull request:
Update limit path creation

The limit path had unnecessary checks for volatility, so in cases of
creating limit over `General` or `SegmentGeneral` locuses (replicated
tables or `generate_series()` for example) limits with just constants
wouldn't pass this checks and generate right path. `Replicated` locus
was also a subject of this issue and on top of that threw an error in
situations when limit was being used with volatile functions.

There's no need for limit to have volatile function inside of it
to force execution on one segment. Because results of such limit path
```sql
->  Limit
    ->  Seq Scan on t_repl
```
is undetermined in case of execution on each segment (as there is
no guarantees on how seqscan will return it's rows) with constant
expression also.

While applying new logic of limit in correlated subplan of `RTE_RESULT`
relation, the parameters were being moved through motions, which lead to
an error. This was happening because path creation for trivial case with single
`RTE_RESULT` relation didn't account situation of having correlated subplan,
like it is accounted for other cases (for example, `RTE_RELATION`). So this
patch is also fixing that.

Co-authored-by: Viktor Kurilko <v.kurilko@arenadata.io>

KnightMurloc commented on src/test/regress/expected/rpt.out:
```diff
@@ -1659,9 +1707,21 @@ select count(distinct(rand)) from cte join d on cte.a = d.a;
 
--- prohibit adding motion on late stage when subplan has external parameters
+-- ensure we're able to create plan when subplan has external parameters
 explain (costs off, verbose) select (select b + random() o) from d;
```
It seems to me that this test has become not what was originally expected.

KnightMurloc commented on src/backend/optimizer/plan/planmain.c:
```diff
@@ -143,8 +143,40 @@ query_planner(PlannerInfo *root,
+				}
+				else 
+					add_path(final_rel, result_path);
+					
```
Why do we need this code?

KnightMurloc commented on src/backend/optimizer/util/pathnode.c:
```diff
@@ -5585,10 +5585,7 @@ create_limit_path(PlannerInfo *root, RelOptInfo *rel,
-		return turn_volatile_seggen_to_singleqe(root, (Path *) pathnode, NULL);
-	else
-		return (Path *)pathnode;
+	return turn_volatile_seggen_to_singleqe(root, (Path *) pathnode, NULL);
```
Update comment.

KnightMurloc commented on src/test/regress/sql/bfv_planner.sql:
```diff
@@ -313,7 +313,7 @@ explain (costs off) select * from t_hashdist cross join (select random() as k, s
 
 -- limit
-explain (costs off) select * from t_hashdist cross join (select * from generate_series(1, 10) limit random()) x;
+explain (costs off) select * from t_hashdist cross join (select * from generate_series(1, 10) limit 1) x;
```
Changing a test for no particular reason is not the good idea.

ShapatinAA commented on src/test/regress/expected/join_hash.out:
```diff
@@ -1013,38 +1013,38 @@ WHERE
-                     Output: hjtest_1.a, hjtest_1.tableoid, hjtest_1.id, hjtest_1.b
-                     Filter: ((SubPlan 4) < 50)
-                     SubPlan 4
+               Output: hjtest_2.a, hjtest_2.tableoid, hjtest_2.b, ((SubPlan 1)), ((SubPlan 3))
```
After some time looking into it, I came to the conclusion that new plan is also a right one.

ShapatinAA commented on src/backend/cdb/cdbpath.c:
```diff
@@ -2774,6 +2774,19 @@ turn_volatile_seggen_to_singleqe(PlannerInfo *root, Path *path, Node *node)
+		 * If we have only LimitPath, without any volatile functions we can
+		 * omit creation of projection path, because it would serve no purpose.
+		 */
+		if (IsA(path, LimitPath) && !contain_volatile_functions(node)) {
```
fixed

ShapatinAA commented:
> Why were another tests changed? 

`join_hash.out`:
While solving the main issue, I faced other problem, which I described [here](https://github.com/GreengageDB/greengage/pull/105#discussion_r2499287400) and was forced to also solve.
This solution changed output of some tests in `join_hash.sql` because they previously went through this logic of path creation for `RTE_RESULT` type relations.

`qp_full_join.out`:
Main issue my patch was solving is forcing `SingleQE` locus after applying `Limit` on nodes with `General` and `SegmentGeneral` locuses.
And in this case we worked on one single segment the whole time because before patch there were `SegmentGeneral` locuses after each of the `Limit`s and so we could do it on one segment and then bring it to dispatcher. But after my patch there is SingleQE locuses on motions to bring it to the place it needs to be after `Limit` nodes. So it firstly gets us to dispatcher and then does it's things. 
As I see it: now we're having also right plan. Maybe less efficient, but definitely more correct in sense of what is going on under the hood (with locuses).

`rpt.out`:
Here we had test, that checked that we also finalize subplans by going recursively in them. After my fix, which I mentioned, again, [here](https://github.com/GreengageDB/greengage/pull/105#discussion_r2499287400), this test was not liquid, as it went through creation of `RTE_RESULT` path also. So to cover, what it was covering new test was added.


> Where is test for root case from the ticket? 

I added this ones which I assume are equivalent in terms of code/case coverage 
```sql
explain (costs off) select * from t_hashdist cross join (select * from t_replicate_volatile limit 1) x;
explain (costs off) select * from t_hashdist cross join (select * from generate_series(1, 10) limit 1) x;
```


> Please, add empty line to the desctiption between titile and commit message.

Fixed

ShapatinAA commented on src/backend/optimizer/plan/planmain.c:
```diff
@@ -143,8 +143,40 @@ query_planner(PlannerInfo *root,
+					add_path(final_rel, path);
+				}
+				else 
+					add_path(final_rel, result_path);
```
Updated this part and added null check.

KnightMurloc commented on src/backend/optimizer/plan/planmain.c:
```diff
@@ -143,8 +143,34 @@ query_planner(PlannerInfo *root,
+						add_path(final_rel, path);
+				}
+				else 
+					add_path(final_rel, result_path);
```
Can we use a variable for the path and make one `add_path` call?

KnightMurloc commented on src/backend/optimizer/util/pathnode.c:
```diff
@@ -5582,13 +5582,12 @@ create_limit_path(PlannerInfo *root, RelOptInfo *rel,
+	 * The limit path's locus must be set to singleQE to ensure deterministic
+	 * ordering of tuples. When distributed across multiple segments, there
+	 * is no guarantee of the global ordering of tuples passed to the limit
+	 * operator, which could yield non-deterministic results.
```
Make it clear that this is only needed for General and SegmentGeneral

ShapatinAA replied:
There is description for `turn_volatile_seggen_to_singleqe()` function in which it is mentioned, that this call is only applicable for `General` and `SegmentGeneral`. Do we need to repeat it there?
Also I think that this description in `turn_volatile_seggen_to_singleqe()`should be corrected according to this patch.

KnightMurloc commented on src/test/regress/expected/rpt.out:
```diff
@@ -1052,6 +1062,67 @@ explain (costs off) select * from t_hashdist cross join (select * from t_replica
  Optimizer: Postgres query optimizer
 (9 rows)
 
+explain (costs off) select * from t_hashdist cross join (select * from t_replicate_volatile limit 1) x;
```
It seems to me that it is better to use inner join if it does not affect the test. Since in practice this is a more common case.

ShapatinAA commented on src/backend/optimizer/plan/planmain.c:
```diff
@@ -143,8 +143,37 @@ query_planner(PlannerInfo *root,
+				{
+					CdbPathLocus outerquery_locus;
+					CdbPathLocus_MakeOuterQuery(&outerquery_locus);
+					path = cdbpath_create_motion_path(root,
```
We could possible lose result path by getting null? Or do you think that this is impossible in this section
